Feat(Certificate): 자격증 전체 조회, 자격증 정보 조회 API 구현 (SWM-49)

morenow98 opened the pull request:
## PR 변경된 내용
### Exam, Subject Entity 클래스 생성 및 매핑
- 두 클래스 모두 Certificate와 다대일 관계로 매핑 (@ManyToOne)
- 테이블에서 외래키 제약조건을 걸지 않기 위해 `@ForeignKey(value = ConstraintMode.NO_CONSTRAINT)` 설정
- Certificate는 Exam이나 Subject 없이도 생성 가능하도록 exams, subjects는 빈 리스트로 필드에서 초기화
- Exam과 Subject 생성 시엔 Certificate가 항상 필요하고, 인자로 들어온 Ceritificate Entity의 exams 혹은 subjects에 자신을 추가하는 로직 구현
### 자격증 전체, 상세 조회 구현
- api 모듈의 usecase에서 @Transactional을 사용하기 위해 api 모듈에 jpa 의존성 추가
- certificate 조회 시 exam과 subject를 지연로딩하고, @BatchSize 를 통해 한 번에 많은 데이터를 불러와 N+1 문제를 해결
- 자격증 당 시험은 100개 이하로, 과목은 10개 이하로 가정해서 BatchSize는 각각 100, 10으로 설정
### 테스트 코드 BDD Mockito 활용해 구현
- 자격증과 관련된 Certificate, Exam, Subject Entity를 만드는 Fixture 클래스 구현
- 아래와 같은 형식으로 작성
```
@DisplayName("자격증 정보 조회 테스트")
@ExtendWith(MockitoExtension.class)
class FindCertificateUseCaseTest {

    @InjectMocks
    FindCertificateUseCase sut;

    @Mock
    CertificateRepository certificateRepository;

    @Test
    @DisplayName("자격증 정보 조회")
    void given_when_then() {
        //given

        //when

        //then
    }

}

```

## 추가 내용
- certificate 관련 API 명세 수정 ("/certificate" -> "/certificates")
- 사용하지 않던 CertificatePersistenceException 클래스 삭제

## 참조
Jira Issue: SWM-49

morenow98 committed:
Feat(Certificate): Exam, Subject Entity 생성 및 매핑

- 외래키로 Unique 제약조건이 걸리지 않도록 설정

morenow98 committed:
Feat(*): api 모듈의 usecase에서 @Transactional 을 사용하기 위해 api 모듈에 jpa 의존성 추가

morenow98 committed:
Feat(Certificate): 자격증 전체, 상세 조회 구현

- Exam 필드명 오류 수정
- API 명세에 s를 붙여 certificates로 수정
- Certificate에서 Exam과 Subject를 각각 1:N 으로 참조
- @BatchSize와 지연로딩을 통해 N+1 문제 해결
- @Transactional을 UseCase 클래스에서 사용

morenow98 committed:
Refactor(Certificate): 사용하지 않던 CertificatePersistenceException 클래스 삭제

morenow98 committed:
Feat(Certificate): Certificate, Exam, Subject Entity에 생성자 추가

- 생성은 정적 팩토리 메서드를 이용하고 생성자를 private으로 설정
- Certificate의 exams, subjects에는 빈 리스트로 생성되도록 설정
- Exam과 Subject 생성 시 인자로 들어온 Certificate의 필드 리스트에 추가

morenow98 committed:
Test(Certificate): 자격증 목록 조회, 자격증 정보 조회 API의 테스트 코드 작성

- 자격증과 관련된 Certificate, Exam, Subject 클래스를 만드는 Fixture 클래스 구현

InHyeok-J commented on jabiseo-api/src/main/java/com/jabiseo/certificate/dto/ExamDto.java:
```diff
@@ -1,10 +1,12 @@
 
+import com.jabiseo.certificate.domain.Exam;
+
 public record ExamDto(
         String examId,
         String description
 ) {
```
Dto와 Response의 네이밍 차이는 뭔가요?

morenow98 replied:
Response는 직접 응답하는 DTO에만 붙였고 Response DTO 안에 부차적으로 들어있는 DTO는 Dto 라고 네이밍했습니다.
적절한 네이밍이 떠오르지 않아서 일단 이렇게 했고, 나중에 바꿔도 좋을 것 같아요

InHyeok-J replied:
개인적으로 reqeust / response 라는 DTO 네이밍을 좋아합니다 
해당 객체가 요청과 응답에 관련있다고 범위를 제한하는것 같아서요 
만약 Exam이 다른 응답에도 포함이 되면 응답의 책임을 같이 공유한다라고 생각해서 Response로 통일시켜도 좋을것 같아요

morenow98 replied:
동감합니다 수정해서 올릴게요

InHyeok-J commented on jabiseo-api/src/main/java/com/jabiseo/certificate/usecase/FindCertificatesUseCase.java:
```diff
@@ -4,10 +4,12 @@
 @Service
+@Transactional(readOnly = true)
 @RequiredArgsConstructor
 public class FindCertificatesUseCase {
```
`FindCertificatesUseCase` 해당 UseCase의 네이밍이 UseCase를 잘 표현하지 못하는것 같습니다... 웬만하면 네이밍으로 댓글 안달려고 햇는데 보기가 너무 어렵고 Find단건 조회랑 코드 차이가 있는거 같아 API 문서를 몇번 봤네요

단건 조회 API -> 이건 자격증의 이름 뿐 아니라 시험,과목 정보를 같이 리턴해야 하고
자격증 List조회 API -> 자격증 이름 목록만 리턴해야 하네요
제 생각에 단건 조회 API가 FindCertificateUseCase <- 해당 Find가 너무 애매한 표현인것 같습니다. 
여러 자격증 목록 조회 -> FindCertificatesUseCsae <- 이렇게 문장이 길때 `s` 만 붙여서 목록을 표현하는건 너무 가혹한 행위라고 생각이들어요 .. 
이러다 보니 dto 패키지에서도 비슷한 클래스가 2개가 있는것 처럼 헷갈리게 만드네요

morenow98 replied:
맞아요 애매한 감이 있었는데,,,
그럼 FindCertificateNameListUseCase 혹은 FindCertificateListUseCsae 로 바꾸는 것도 괜찮을 것 같아요
이것 말고도 더 괜찮은 네이밍이 있을까요?

InHyeok-J replied:
세상에서 가장 어려운걸 물어보네요
FindCertificateUseCase-> FindCertificateDetailUseCase
FindCertificatesUseCsae -> FindCertificateListUseCsae 이정도로 할까요?

morenow98 replied:
좋습니다 수정해서 올릴게요

InHyeok-J commented on jabiseo-domain/src/main/java/com/jabiseo/certificate/domain/Certificate.java:
```diff
@@ -22,10 +22,29 @@ public class Certificate {
+    @OneToMany(mappedBy = "certificate")
+    private List<Subject> subjects = new ArrayList<>();
+
+    private Certificate(String id, String name) {
         this.id = id;
         this.name = name;
     }
 
+    public static Certificate of(String id, String name) {
+        return new Certificate(id, name);
+    }
```
생성자를 private로 바꾸고 정적 팩토리 메소드를 적용한 이유가 있을까요? 이건 궁금하네요

morenow98 replied:
엘레강트 오브젝트 에서는 생성자가 로직을 수행하지 않아야 한다고 합니다.
생성자는 인자로 들어온 것들로 객체를 만들어내는 일 이외에는 아무것도 하지 않아야 할 투명한 존재로 남겨두어야 예측이 편하고 테스트도 쉽다는 이유때문에요.

하지만 Exam과 Subject Entity 클래스에서는 인자로 넘어온 certificate로 Entity를 만들기만 하면 문제가 발생합니다. certificate의 exams 혹은 subjects 에 자신이 없을 수 있기 때문입니다.
1번 자격증에 2번 과목을 만들었다고 가정해보면, 2번 과목에는 1번 자격증이 있을겁니다. 하지만 과목을 만들 당시 1번 자격증의 과목 리스트를 수정하지 않았다면 자격증의 과목 리스트에는 2번 과목이 존재하지 않게 됩니다.
따라서 과목을 만들 때 인자로 넘어온 자격증도 같이 수정되어야 한다고 생각했고, 이를 위해선 생성자에 로직이 들어가야 했습니다.

하지만 처음에 말한 이유때문에 생성자에 직접 로직을 넣는 게 꺼림칙하게 느껴졌습니다. 따라서 생성자는 투명하게 남겨놓되 정적 팩토리 메소드로 로직을 구현했습니다. 또한 생성자에 직접 접근한다면 위에 말한 문제점이 해결되지 않아 생성자는 private으로 수정했습니다.

방금 말한 문제는 Exam과 Subject 두 클래스에 해당하는 내용이지만, Certificate도 통일해서 정적 팩토리 메소드만 사용해 생성하도록 하는 것이 좋아보여 Certificate에도 같은 방식을 적용했습니다.

InHyeok-J replied:
좋은 내용이네요. 도메인 객체에서 생성자 가 로직을 수행하지 않는다는건 좋은 내용같습니다 (물론 예외는 있습니다)

하지만 Certificate에 통일되면 나중에 필드 값이 많은 경우 builder를 써야 하는데 관례상 통일시키면 못쓰는 상황이 발생하지 않을까요?

InHyeok-J replied:
물론 그때그때 상황봐서 public하게 열어도 괜찮으면 상관없습니다~

morenow98 replied:
저는 아래와 같은 이유때문에 builder를 쓰는 것을 선호하지 않습니다.

- A클래스에 a, b 필드가 있어서 A.builder.a(a).b(b).build() 와 같이 A클래스를 생성하는 상황.
- A클래스에 c 필드를 추가한 후 위와 같은 코드에서는 에러가 발생하지 않고 c에 null이 들어가 제어하기 힘듦.
- 만약 new A(a, b) 와 같이 생성하거나 A.of(a, b) 와 같이 생성했다면 컴파일 에러가 발생해 c가 누락될 수 없음.

따라서 정적 팩토리 메소드 혹은 생성자만을 사용해 클래스를 생성하는 것을 선호합니다.
Builder를 사용하면 어느 필드에 어떤 값이 들어가는지 명확하게 알 수 있다는 장점이 있지만, 요즘 IDEA에선 어느 필드에 어떤 값이 들어가야 하는지 하나하나 알려주는 기능이 있어 이런 장점도 많이 퇴색되었다고 알고 있습니다.

하지만 Builder를 사용했을 때 장점이 있을 수 있으므로 추후에 통일성을 포기하고 생성자를 public으로 전환해 Builder 패턴을 사용할 수 있다고 생각합니다.

InHyeok-J replied:
동의합니다~
null 체크는 builder의 문제가 되죠 이건 상황에 따라 선택하는게 좋아보이네요 
또 정적 팩토리 메소드 패턴에서 가장 큰 장점 중 하나는 메소드 이름을 지정할 수 있다고 생각이 드는데요
개인적으로 저 `of`라는 메소드 이름이 도메인 객체의 생성 행위를 나타낼 수 없다고 생각합니다. 
이팩티브 자바에서는 ~~~ 값들로 객체를 생성한다 이지만, 도메인 객체의 생성에는 네이밍이 명확하게 처리되어야 추후 다른 스태틱 메소드를 추가할 수 있다고 생각이 드는데 이 부분에 대해서는 어떻게 생각하세요?

morenow98 replied:
너무 좋은 것 같아요. 저는 평소에 정적 팩토리 메서드를 사용할 때 그냥 of나 from 위주로 사용하고 다른 네이밍은 필요할 때만 달아서 사용했던 것 같아요.

특히 Exam, Subject는 메서드 안에서 Certificate 리스트에 추가되는 로직이 있어서 그 로직을 메서드에 표현해보려고 했지만, 명확한 메서드명이 떠오르지 않더라구요. 좋은 메서드 명을 생각해서 바꾸면 좋을 것 같아요!

morenow98 committed:
Fix(Certificate): Certificate관련 에러를 정확하게 던지지 않던 오류 수정

morenow98 committed:
Refactor(Certificate): 자격증 조회 usecase와 dto 이름을 명확하게 수정

- FindCertificate -> FindCertificateDetail
- FindCertificates -> FindCertificateList

morenow98 committed:
Refactor(Certificate): Dto 라는 이름의 dto들을 명확한 의미 전달을 위해 request/response로 변경

- 기존 Dto 클래스들은 직접 응답되지는 않는, dto들의 안에 포함되어 있던 부차적인 dto 클래스들이었지만, request/response 중 관여하는 쪽으로 이름을 붙이기로 결정
